### ISSUE TITLE
Incorrect calculation of multiple sales tax

### ISSUE THREAD

**Issue by meadlin** [jekkos/opensourcepos#1163]
### IMPORTANT IMPORTANT IMPORTANT

Chose what you want to report: New Feature or Bug.
If you remove the template when submitting a Bug your issue will be closed as we cannot help without basic information.

### New Feature / Enhacement

For New Features or Enhacements please remove all the template text and clearly write your proposal.
It's important to state whether you expect the community to implement it or you will contribute the work.
Please bear in mind that we will implement new features only on the current code, there is no support for old versions.

### Issue / Question / Bug

Before submitting an issue please make sure you remove the first section of the template and you tick (add a x between the square brakets) and agree with all the following check boxes:

- [X] Checked the current issues database and no similar issue was already discussed
- [X] Read the README, WHATS_NEW and UPGRADE
- [X] Read the FAQ (https://github.com/jekkos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP has php5-gd, php-intl and sockets installed)
- [X] Reporting an issue of an unmodified OSPOS installation
- [X] PHP version is at least 5.5 and not 7.x
- [X] MySQL version is 5.5 or 5.6 and not 5.7
- [X] Ran any database upgrade scripts (e.g. database/2.4_to_3.0.sql)
- [X] Know the version of OSPOS and git commit hash (check the footer of your OSPOS) and will add to my issue report
- [X] Know the name and version of OS, Web server and MySQL and will add to my issue report

### Installation information
OSPOS:  3.1.0 - 4f5ad57
OS: Arch linux
PHP:  7.0.14-1
MySQL: (Mariadb: 10.1.21-1)

### Expected behaviour
When multiple tax districts are present, calculate and round the sales tax of each instead of as a sum of the taxes.

### Actual behaviour
Multiple sales tax percentages are summed, then the tax is calculated/rounded from the total. The individual sales taxes are displayed correctly on the receipt, but the total is generated as a result of the total tax instead of the individual taxes being summed.

### Steps to reproduce the issue
Sales tax 1: 4.75%
Sales tax 2: 3.75%

subtotal: $450
tax 1: $21.38
tax 2: $16.88
Total: $488.25

Which causes an accounting deficit of $0.01 as the total should be $488.26.

**Comment by daN4cat** [comment]
Many thanks for the explanation, this should be the recommended fix for #1094.

At the moment the steps are to calculate the sum of the two taxes % and then apply it to the subtotal to obtain the total, rounding the total.

So each tax % needs to be applied to the subtotal, get the rounded tax and then sum the two taxes on top of subtotal to obtain the total.

I believe there will be some work to change this.

**Comment by SteveIreland** [comment]
Thanks for addressing this @daN4cat .

I've been working on the sales tax module for awhile now.  It's expanding as I learn the nuances of the various tax legislation of each state (at least the ones I have to immediately account for).

But I want to keep an open mind toward VAT tax because there is discussion going around suggesting that the big Donald might trump us with a VAT tax.  So in the example above is this a  pure Sales tax or a VAT tax or a combination of both?  I always assumed that everyone else except for the U.S. used VAT taxes but I'm thinking that I might be totally wrong on this point.

**Comment by meadlin** [comment]
This example is for pure sales tax. 4.75% for the state and 3.75% for the city, both require their sales taxes be rounded up.

**Comment by daN4cat** [comment]
@SteveIreland in EU/UK prices are with VAT included. That's why @jekkos and I tend to use and test with that.
This two level of taxation is mainly a US thing, and it's tricky to deal with when calculating totals and etc.

I've just checked Sale model and the searches need to change completely to move from sum of tax % to the correct way.
And to be honest I'm not so inspired because of that work I would use 0. Oh joy...

**Comment by meadlin** [comment]
I'll take a look at it and the rest of the code to see if I can come up to speed and help take a stab at it. It might be beneficial to add the capability to add more than two tax districts, as in my example at least, the state is really 4 different sales tax districts that are being combined to conform to the SW model.

**Comment by daN4cat** [comment]
@meadlin it's all in models/sale.php
I'm currently checking what can be done...

**Comment by daN4cat** [comment]
I'm currently working on this....

**Comment by daN4cat** [comment]
... and I fixed the summary report as first step.

**Comment by daN4cat** [comment]
@meadlin @odiea see my last commit to master. Please test as I reworked the tax calculation.
Hopefully I managed to fix all was required, but you never know.... it took half day of work!

**Comment by meadlin** [comment]
@daN4cat Just pulled, and it looks like the total still isn't updating correctly.

![salestax](https://cloud.githubusercontent.com/assets/22114210/23347494/f54abff2-fc5f-11e6-8070-dda1e7ba5972.png)

**Comment by daN4cat** [comment]
@meadlin that is with the register only or even the receipt?

**Comment by meadlin** [comment]
@daN4cat both, after adding a payment and printing, still came out short.

**Comment by meadlin** [comment]
@daN4cat Actually, there doesn't seem to be a way to complete the transaction now.

**Comment by daN4cat** [comment]
Ok

**Comment by odiea** [comment]
Now it appears that reports are showing that lost penny Ok Bust Register and Receipts are incorrect.
![2017-02-27 05_08_00-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/23360936/48f7bbc8-fcab-11e6-95f2-27c2e7cebf4d.png)
![2017-02-27 05_08_25-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/23360947/574c2682-fcab-11e6-885e-aab504e8f1c2.png)

**Comment by meadlin** [comment]
@odiea On a side note, how did you git the third sales tax? I can't seem to get more than two.

**Comment by odiea** [comment]
I had 1 item at 7.4 and the second item at 3.25 and 7.35 so it was 2 items sold.

**Comment by meadlin** [comment]
@odiea Ok, I was thinking it was 3 different tax rates on the same item(s).

**Comment by odiea** [comment]
No here in Ft fun we just use the combine sales tax .

**Comment by meadlin** [comment]
@daN4cat Thanks, nice to know. It'll definitely help with accounting/reporting.

**Comment by daN4cat** [comment]
@meadlin @odiea if you reprint a receipt do you get the same issue or no?

**Comment by daN4cat** [comment]
because when in sale register mode the tax calculation is done in Sale_lib.php which is different from the reports or re-print of a receipt.... weird isn't it?

**Comment by meadlin** [comment]
@daN4cat Reprint has the same error.

**Comment by daN4cat** [comment]
ok, thanks

**Comment by odiea** [comment]
Yes the receipt I showed was from a reprint.

**Comment by daN4cat** [comment]
So the problem is the total, taxes are correct now.

**Comment by meadlin** [comment]
@daN4cat Yea, from what I've seen, the taxes have been correct, just the total was the issue.

**Comment by odiea** [comment]
Yes. With a calculator they were showing the correct value. It depends on the rounding I guess.

**Comment by daN4cat** [comment]
Yes, trying to figure out were it goes wrong with the total...

**Comment by daN4cat** [comment]
And when in detailed sale, having just one line is the issue the same or no? I guess it's the same.

**Comment by meadlin** [comment]
The issue, for me at least, is that the taxes to be collected are the sum of the rounded district taxes, while the total appears to have the sum of the district tax rates applied to it.

I.e.:
subtotal + ROUND(subtotal*tax1) + ROUND(subtotal*tax2)
vs
subtotal + ROUND(subtotal* (tax1+tax2))

**Comment by daN4cat** [comment]
could be the rounding to 3 decimals instead of 2... because both values are something like 23.375 + 12.325

if they are rounded at display time the look ok, but in the total they'll make -0.01

**Comment by meadlin** [comment]
Possibly

**Comment by daN4cat** [comment]
That part of the code related to receipt and invoice is convoluted. It uses a different branch and redoes the tax calculations in a different way from what is fetched from the database in the reports/tabular views.

I fixed the latter not the receipt/invoice. Yes the code is really crap in this part...

**Comment by odiea** [comment]
Quite a mess you inherited.

**Comment by meadlin** [comment]
As is the way of any software project :smile:

**Comment by jekkos** [comment]
I told you that this is one of the hardest parts in this application.. A domain driven design might simplify things here..
It's just that once features are added and nothing refactored technical debt will keep increasing. Think we need to think about this one moment or another.

**Comment by daN4cat** [comment]
It would probably make sense to fetch the cart line from the same temp table style not straight from the db.

This requires some serious refactoring as now the core part is working fine, it's just those silly methods `get_sale_items_ordered` and `get_sale_items` that fetch the data from the DB and then rely on the Sale_lib to redo exactly the same it's done in the other methods leveraging SQL functions....

grrrrrrrrrrrr

**Comment by jekkos** [comment]
Hm ic, I think it has always been that way, I spent quite some time in the past in that part of the receipt code. But it's annoying that there are two ways. 

On the other hand I think it's a pity that CI has such poor testing support. If we'd want to refactor this part, I'd prefer to write some tests first so one can go ahead without breaking half of the application

**Comment by daN4cat** [comment]
@meadlin @odiea please check the latest master and see if this fix solves the problem.

**Comment by daN4cat** [comment]
I rounded the tax in Sale_lib `get_taxes` function as that was giving the output with no rounding so 3 or 4 decimals.
The mistake in your example was with 3 decimals and the last one being 5 on both, so adding up to the 0.01 difference.

**Comment by odiea** [comment]
Looks good to me now. Both register and Receipt is adding up correctly.
![2017-02-27 16_50_40-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/23385446/1d3ebd9e-fd0d-11e6-96d2-030598f25c8c.png)

**Comment by meadlin** [comment]
When I test the change with a new sale, the taxes/total come out correct. 
![newsale](https://cloud.githubusercontent.com/assets/22114210/23386312/ec8513fa-fd12-11e6-9859-93aa4c00b960.png)

When I pulled up a historical sale, the taxes are now rounded down, but the total is correct for the false rounding.
![historicalsale](https://cloud.githubusercontent.com/assets/22114210/23386319/f3c3198c-fd12-11e6-9612-606b6b19b194.png)

The takings for the new sale also still show the rounded down value
![takingsreport](https://cloud.githubusercontent.com/assets/22114210/23386326/03088e5e-fd13-11e6-9581-3ab3698e49e4.png)

**Comment by daN4cat** [comment]
Ok, actual is now fine and historical wrong. Do you see the same with the reports? I mean detailed an summary?

**Comment by meadlin** [comment]
Detailed 
![incorrectdetailed](https://cloud.githubusercontent.com/assets/22114210/23393661/aab27290-fd41-11e6-8381-9629c7ce3215.png)
and summary
![incorrectsummary](https://cloud.githubusercontent.com/assets/22114210/23393670/b768833a-fd41-11e6-8b61-f72d72e59a11.png)

are both incorrect now, oh the joys of multi-sales tax...

**Comment by daN4cat** [comment]
Yep, big issue....

**Comment by daN4cat** [comment]
I wonder what it is now that makes the stored data go wrong.
I guess in fixing the round on the cart line fetch calculation I interfered with a flow in the code related to saving the data.
That can be the only explanation as the totals seens now consistent just the taxation is stored not right.

**Comment by meadlin** [comment]
It may be stored correctly, those reports were for the original historic sale. I'm trying to generate a new sale now to test. Anyone know why on occasion after adding a payment, the complete button doesn't appear?

**Comment by meadlin** [comment]
Got a new sale to work, takings, summary and detailed are all still incorrect.

**Comment by RamkrishnaMondal** [comment]
@meadlin  Can you set tax decimal to 3 and check ?

**Comment by meadlin** [comment]
@RamkrishnaMondal I cannot test anything using new data at the moment. Sadly I cannot complete a sale, or suspend one for that matter....I don't know how I got it to work last night, but I'm dead in the water so far. Hopefully I can get it working without too much trouble.

**Comment by meadlin** [comment]
ok, had to update the DB, to get going again... Weee...anyways, with tax decimal set to 3, detailed/summary reports and takings are still incorrect.

**Comment by daN4cat** [comment]
The issue is the 0.01 issue that is outside the rounding interval that would enable the complete button.

**Comment by daN4cat** [comment]
Add a payment of 0.005 and you'll see it enabled.

**Comment by araalrai** [comment]
Another issue is with when the "Tax Included" is enabled. It does not calculating taxes properly when you have two taxes. For example:

Tax 1 @ 5%
Tax 2 @ 7%

Lets say the total price including taxes is $112
The break down should be as follows:
Sub Total: $99.34
Tax1: $5.33 
Tax2: $7.33

It should be:
Sub total: $100
Tax1 $5
Tax2 $7

What the program is doing is basically calculating each tax separately. So it is taking the total $112 x Tax 1 divided by 100 + Tax 1 = $5.33

Any fix or suggestions so i may fix it would be appreciated.

Thanks

**Comment by araalrai** [comment]
Correction to my previous comments:

The breakdown shows as follows:
Sub Total: $99.34
Tax1: $5.33
Tax2: $7.33

It should be:
Sub total: $100
Tax1 $5
Tax2 $7

**Comment by daN4cat** [comment]
@araalrai that problem is part-fixed in current master, however there is an additional issue when it comes to lookup an past sale.

**Comment by araalrai** [comment]
Could you please let me know which php files are used to calculate sales
taxes. I may have something which i may implement which may fix the tax
related issues.

**Comment by daN4cat** [comment]
It's convoluted, you really know well the code to understand where.

Most likely your issue is fixed already and touches 3 files. Just check the commit history to understand which ones. I'm on my mobile phone at the moment and far from the code.

**Comment by SteveIreland** [comment]
FYI - I'm addressing these calculation issues in the "soon to be available" Customer Sales Tax module.

**Comment by daN4cat** [comment]
@SteveIreland if you are reworking this part I don't spend anymore time to investigate and fix bugs.
I'll focus on integrating table pr next days.

**Comment by jekkos** [comment]
@SteveIreland Might be cool if you could try to add some unit tests for this crucial part of the application. Php unit support should be there using composer . Tax calculations feels like something that should be well coherent and modular to write some tests for.

**Comment by jekkos** [comment]
Having unit tests will help us refactor in the future without making regressions.. you add support for one scenario and break a whole couple of others. This is painful and kind of demotivates people to dive into the code

**Comment by SteveIreland** [comment]
I will look into this over the weekend.

**Comment by daN4cat** [comment]
@jekkos it's all the fun with dealing with OSPOS code, it's a free of charge endurance adventure 😊

**Comment by daN4cat** [comment]
#1133 supersedes this issue.

**Comment by daN4cat** [comment]
@odiea is this now fixed in master?

**Comment by odiea** [comment]
It appears that all reports and receipts are showing the same now. Looks good to me.
![2017-04-16 06_17_50-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/25071378/8fa21824-2272-11e7-97f6-007f1909699d.png)
![2017-04-16 06_21_25-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/25071379/97e98972-2272-11e7-9f2c-7c2912026311.png)

**Comment by meadlin** [comment]
What's the easiest way to apply the DB changes without losing data? I can try it with my data set here too.

**Comment by daN4cat** [comment]
Make sure you make a copy of your database.
Use the script in database folder, the one to upgrade from 3.0.2 to 3.1.0. Not sure what is the starting version but you may just need to use of that script the last part after the rewards upgrade to get the changes Steve introduced.

**Comment by meadlin** [comment]
K, thanks

**Comment by meadlin** [comment]
Register/Receipt look good:
![salesreciept](https://cloud.githubusercontent.com/assets/22114210/25077474/72e7a3a0-22e7-11e7-81ea-8016546b670b.png)

Detailed, summary and takings are still off:
![detailedreport](https://cloud.githubusercontent.com/assets/22114210/25077476/824ad3d0-22e7-11e7-94dd-b59a9fba13c2.png)
![summaryreport](https://cloud.githubusercontent.com/assets/22114210/25077478/853a3ffe-22e7-11e7-94d4-2f0a84ea3322.png)
![takings](https://cloud.githubusercontent.com/assets/22114210/25077479/86ca0340-22e7-11e7-86c4-22d42e49144c.png)

Unless I messed something up in the database update.

**Comment by SteveIreland** [comment]
I don't think you messed anything up

The focus of my corrections was on the application of payment to the amount due in the register panel and getting that number to calculate correctly every time (i.e. no rounding errors regardless of the affect of tax rounding ... regardless of the number of tax entities involved.

Reports are a different matter.  A brief exchange suggested that I shouldn't try to provide a data migration script for existing data so I didn't.  And because I didn't provide a data migration script I could not change the SQL for the reports - so the reports are still impacted by potential rounding errors.

I should be capturing the correct tax amount on the sales_taxes table, but reports are still being driven by recalculating the sales tax via the sales_item_taxes table and that still has the possibility of introducing rounding errors.  After about a year I was going to come back and ask to see if there was any interest in migrating the reports over to use the sales_tax tables instead of the sales_item_tax table (since everyone would by then have a good year of sales history).

On a related note, the sales tax is still a single tax rate regardless of the number of tax jurisdictions in play. I haven't seen anything to suggest that this isn't acceptable and I've been working on tax systems for a long time. 

All indications are that no one wants to burden OSPOS with back office tax, payments, expense tracking and profit reporting.  So I backed off of breaking out the tax rates by tax jurisdiction and I'm deferring tax reporting to an external module (which I will start development on this month).  This will break out the sales taxes to collected so that appropriate tax reporting and payment can take place.  Of course, I'll be primarily focused on making it work for my client, but I will also include support for any other state specific requirements for anyone who wants to try it out when it's finished.

**Comment by daN4cat** [comment]
@SteveIreland I was about to write this morning too because I'm really confused.
We agreed that your tax rework will replace all the current tax calculations and have one approach to all.
I'm double confused by your statement about the migration, I said at that time that yes you need to provide a migration step in the database script upgrade.

I'm afraid that if we don't rework totally this tax story throughout the current code, it will be just an half baked solution and I'm not sure it's what we want to release in 3.1.0.

Could you please check the rest of the code and migrate all tax codes and taxes to the new system and have reports working properly?
Reports use one base class and function, so it should be easy to do one fix for all.

I'm fine with the tax calculation story as office backend story.

**Comment by jekkos** [comment]
I also think that we should do it all or nothing. I'd rather stick with a  more crippled but consistent functionality instead of having something that's half done. I do  need to have a look at the code as I'm a bit left behind there atm

**Comment by SteveIreland** [comment]
Okay.  By the end of this week I'll have an option for migrating prior data and modify the tax reporting to use the new sales_tax table.

**Comment by daN4cat** [comment]
@SteveIreland sounds good. Thanks.

**Comment by SteveIreland** [comment]
I'm going to go ahead a split this into two pull requests.  I finally have the sales tax history migration working in a somewhat reasonable way.  I spent far too much time trying to make it elegant and never achieved nirvana.  So I reverted back to something a little more simple, but I think might be acceptable.

I'll submit the pull request for the migration today with an explanation as to how to run it.  It's simple but may not be intuitive.

If the migration meets with majority approval (does it need to be a super majority?) then I'll proceed with the changes to the reporting that will use the new tax data.  I just don't want to break anything before it's ready.
